numcodecs/sz.py

```diff
@@ -0,0 +1,84 @@
+import ctypes
+from numcodecs.abc import Codec
+
+
+class Params(ctypes.Structure):
+
+    _fields_ = [
+        ("options_mask", ctypes.c_int),
+        ("bits_per_pixel", ctypes.c_int),
+        ("pixels_per_block", ctypes.c_int),
+        ("pixels_per_scanline", ctypes.c_int)
+    ]
+
+
+libsz = False
+for ext in [".so", ".dylib", ".dll"]:
+    try:
+        libsz = ctypes.cdll.LoadLibrary("libsz" + ext)
+    except OSError:
+        pass
+
+if libsz:
+    BufftoBuffCompress = libsz.SZ_BufftoBuffCompress
+    BufftoBuffCompress.restype = ctypes.c_int
+    BufftoBuffCompress.argtypes = [
+        ctypes.c_void_p, ctypes.POINTER(ctypes.c_size_t), ctypes.c_void_p, ctypes.c_size_t,
+        ctypes.POINTER(Params)
+    ]
+
+    BufftoBuffDecompress = libsz.SZ_BufftoBuffDecompress
+    BufftoBuffCompress = libsz.SZ_BufftoBuffCompress
+    BufftoBuffCompress.restype = ctypes.c_int
+    BufftoBuffCompress.argtypes = [
+        ctypes.c_void_p, ctypes.POINTER(ctypes.c_size_t), ctypes.c_void_p, ctypes.c_size_t,
+        ctypes.POINTER(Params)
+    ]
+
+
+def check_sz():
+    if not libsz:
+        raise ImportError("libsz could not be loaded")
+
+
+class HdfSzipCodec(Codec):
+
+    codec_id = "hdf_szip"
+
+    def __init__(self, mask, bits_per_pixel, pix_per_block, pix_per_scanline):
+        self.mask = mask
+        self.pix_per_block = pix_per_block
+        self.bits_per_pixel = bits_per_pixel
+        self.pix_per_scanline = pix_per_scanline
+
+    def decode(self, buf, out=None):
+        check_sz()
+        buf = memoryview(buf)
+        param = Params(self.mask, self.bits_per_pixel, self.pix_per_block, self.pix_per_scanline)
+        lout = int.from_bytes(buf[:4], "little")
+        dest_len = ctypes.c_int()
```

Comment:
```suggestion
        dest_len = ctypes.c_int(lout)
```

I think we want to provide the allocated size of the buffer as an input to `BuffertoBuffDecompress`.

Reply:
OK - the comments seems to suggest that this was to be set by the decompressor. 

Still crashes the example once applied.